pkg/controller/metrics.go

```diff
@@ -18,41 +18,323 @@ limitations under the License.
+			"namespace":              msMeta.Namespace,
+			"uid":                    string(msMeta.UID),
+			"available_replicas":     string(machineSet.Status.AvailableReplicas),
+			"fully_labeled_replicas": string(machineSet.Status.FullyLabeledReplicas),
+			"ready_replicas":         string(machineSet.Status.ReadyReplicas),
+			"replicas":               string(machineSet.Status.Replicas)}).Set(float64(machineSet.Status.ReadyReplicas - machineSet.Status.Replicas))
+
+		if machineSet.Status.FailedMachines != nil {
+
+			for _, failed_machine := range *machineSet.Status.FailedMachines {
+				metrics.MachineSetStatusFailedMachines.With(prometheus.Labels{
+					"name":                                msMeta.Labels["name"],
+					"namespace":                           msMeta.Namespace,
+					"uid":                                 string(msMeta.UID),
+					"failed_machine_name":                 failed_machine.Name,
+					"failed_machine_provider_id":          failed_machine.ProviderID,
+					"failed_machine_last_operation_state": string(failed_machine.LastOperation.State),
+					"failed_machine_last_operation_machine_operation_type": string(failed_machine.LastOperation.Type),
+					"failed_machine_owner_ref":                             failed_machine.OwnerRef}).Set(float64(1))
+			}
+		}
+	}
 }
 
-// Collect is method required to implement the prometheus.Collect interface.
-func (c *controller) Collect(ch chan<- prometheus.Metric) {
+// CollectMachines is method to collect Machine related metrics.
+func (c *controller) CollectMachineMetrics(ch chan<- prometheus.Metric) {
 	// Collect the count of machines managed by the mcm.
 	machineList, err := c.machineLister.Machines(c.namespace).List(labels.Everything())
 	if err != nil {
-		ScrapeFailedCounter.With(prometheus.Labels{"kind": "machine-count"}).Inc()
+		metrics.ScrapeFailedCounter.With(prometheus.Labels{"kind": "Machine-count"}).Inc()
 		return
 	}
-	metric, err := prometheus.NewConstMetric(machineCountDesc, prometheus.GaugeValue, float64(len(machineList)))
+
+	for _, machine := range machineList {
+		mMeta := machine.ObjectMeta
+		mType := machine.TypeMeta
+		mSpec := machine.Spec
+
+		metrics.MachineCreated.With(prometheus.Labels{
+			"name":      mMeta.Name,
+			"uid":       string(mMeta.UID),
+			"namespace": mMeta.Namespace}).Set(
+			float64(mMeta.GetCreationTimestamp().Time.Unix()))
+
+		metrics.MachineInfo.With(prometheus.Labels{
+			"name":                 mMeta.Name,
+			"namespace":            mMeta.Namespace,
+			"uid":                  string(mMeta.UID),
+			"generation":           strconv.FormatInt(mMeta.Generation, 10),
+			"kind":                 mType.Kind,
+			"api_version":          mType.APIVersion,
+			"spec_provider_id":     mSpec.ProviderID,
+			"spec_class_api_group": mSpec.Class.APIGroup,
+			"spec_class_kind":      mSpec.Class.Kind,
+			"spec_class_name":      mSpec.Class.Name}).Set(float64(1))
+
+		for _, condition := range machine.Status.Conditions {
```

Comment:
The conditions are handled same for `Machines`, `MachineSets` and `MachineDeployments`. Could we move that into a separate function to which we pass the metric, label information and the array of conditions. Then we do not need to duplicate code.

Reply:
I tried that, but because all condition types are different structs (also not interfaced) it is not working. I tried accepting them as interface{} , but this also leads to a non-iterable. 

The Status fields are not the same type. 1 of them is from k8s api, 2 of them are from mcm api. 

So I don't think that's a good idea to invest much time into this.